src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs

```diff
@@ -112,60 +115,106 @@ public virtual async Task<ProjectPackages> GetInstalledAndTransitivePackagesAsyn
-                InstalledPackages.Clear();
-                TransitivePackages.Clear();
-                targetsList = await GetTargetsListAsync(assetsPath, token);
+                installedPackages = new T();
+                transitivePackages = new T();
+                targetsList = (await GetTargetsListAsync(assetsPath, token))?.ToList();
+            }
+            else
+            {
+                if (InstalledPackages == null || TransitivePackages == null)
+                {
+                    installedPackages = new T();
+                    transitivePackages = new T();
+                }
+                else
+                {
+                    // Don't mutate cache for threadsafety, instead we works on copy then replace cache when done.
+                    lock (_lockObj)
+                    {
+                        (installedPackages, transitivePackages) = GetCacheCopy();
+                    }
+                }
             }
 
             var frameworkSorter = new NuGetFrameworkSorter();
-
             // get installed packages
-            List<PackageReference> installedPackages = packageSpec
+            List<(IReadOnlyList<PackageReference> packageReferences, FrameworkInstalledPackages detectedInstalledPackageChanges)> fetchedInstalledPackagesList = packageSpec
                 .TargetFrameworks
-                .SelectMany(f => FetchInstalledPackagesList(f.Dependencies, f.FrameworkName, targetsList))
+                .Select(f => FetchInstalledPackagesList(f.Dependencies, f.FrameworkName, targetsList, installedPackages)).ToList();
+
+            // Update installedPackages only after fetching is done for thread-safety.
+            UpdatePackageListWithNewPackageIdsAndApplyNewVersions(installedPackages, fetchedInstalledPackagesList.Select(f => f.detectedInstalledPackageChanges));
+
+            List<PackageReference> calculatedInstalledPackages = fetchedInstalledPackagesList
+                .SelectMany(f => f.packageReferences)
                 .GroupBy(p => p.PackageIdentity)
                 .Select(g => g.OrderBy(p => p.TargetFramework, frameworkSorter).First())
                 .ToList();
 
             // get transitive packages
-            IEnumerable<PackageReference> transitivePackages = packageSpec
+            List<(IReadOnlyList<PackageReference> packageReferences, FrameworkInstalledPackages detectedTransitivePackageChange)> fetchedTransitivePackagesList = packageSpec
                 .TargetFrameworks
-                .SelectMany(f => FetchTransitivePackagesList(f.FrameworkName, targetsList))
+                .Select(f => FetchTransitivePackagesList(f.FrameworkName, targetsList, installedPackages, transitivePackages)).ToList();
+
+            // Update transitivePackages only after fetching is done for thread-safety.
+            UpdatePackageListWithNewPackageIdsAndApplyNewVersions(transitivePackages, fetchedTransitivePackagesList.Select(f => f.detectedTransitivePackageChange));
```

Comment:
If the above is operating on a copy why do we need to update it again?

Reply:
Well, I could have mutated my input `transitivePackages` while processing, but it's still my input, mutating input may introduce some unwanted feedback loop. I'm trying to set good practice (ideally, pure functions) so next time don't have any subtle bug. I believe mutating your input while processing then looping through is not good practice, very hard to reason, if there is any mutation need then apply after finish processing current input.

Reply:
but the transitivePackages are already mutated anyways no? 
The method used above hasn't changed that.

Reply:
Yes, only 1 case for now, it's in CpsPackageReferencesProject.cs, but not in LegacyPackage. 

https://github.com/NuGet/NuGet.Client/pull/4498/files#diff-5fae96e8312674fe53149dbdb5a5f036aea82698a5aeff49617e8e7ae66d2685R248

I'm trying to make it close to pure function as much as possible but couldn't do in 1 PR without breaking some logics since code is very tangled.

Reply:
There is another [global cache](https://github.com/NuGet/NuGet.Client/blob/df0937651cd594fe999f6a2ee4c67338ca4b17b8/src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs#L39) need to be addressed before `Transitive Package` goes alive. Maybe it that [mutation](https://github.com/NuGet/NuGet.Client/pull/4498/files#diff-5fae96e8312674fe53149dbdb5a5f036aea82698a5aeff49617e8e7ae66d2685R248) can be addressed with it.

Reply:
I think "pure function" is kind of misleading.
The mutating approach is more allocation friendly than creating collections consistently. It's all fair game as long as it's obvious what it does (docs, naming, etc).

Just because linq is used it doesn't mean it has to be used in a functional programming way.

Reply:
So going back to it, wouldn't removing UpdatePackageListWithNewPackageIdsAndApplyNewVersions be a no-op?

Reply:
I thought current approach it would make code more future proof and cleaner in long run. I want to untangle it since I had chance to go through code. If people keep adding more logics here (I know more transitive dep changes coming), then subtle bug like can reintroduced unless we change approach here.
I believe correctness is more important than memory allocations [considering there're so many hits](https://github.com/NuGet/Client.Engineering/issues/1500).

I'm not sure if removing `UpdatePackageListWithNewPackageIdsAndApplyNewVersions` wouldn't change correctness, I can check again.

Reply:
> I believe correctness is more important than memory allocations https://github.com/NuGet/Client.Engineering/issues/1500.

I agree correctness is what matters, but you are implying that the current approach has no way of being correct, which isn't true. 
It's just not thread safe.

Reply:
I removed it. Please check latest code.